### ISSUE TITLE
Query with errorPolicy "all" cannot use onError, onCompleted effectively

### ISSUE THREAD

**Issue by tdeedlesATX** [apollographql/apollo-client#6966]
Hi 👋 ! A team of us are working on transitioning a production application to Apollo with use of hooks in React. We have the following use case:

- use a query with errorPolicy set to "all" (so that `error` and `data` may both have value)
- use `onError`, `onCompleted` attributes to perform some side effects

## Intended outcome:

I'd expect to be able to perform "completed" and "errored" functionality if the errorPolicy="all", since this would be a congruent outcome of the `data, error, loading` pattern commonly used when both `data` and `error` may be populated with this error policy.

## Actual outcome:

If there is a partial error from one of the fields, only `onError` called (since I've defined both).
I think issue can be found in this either/or method ([here](https://github.com/apollographql/react-apollo/blob/v2.5.8/packages/hooks/src/data/QueryData.ts#L329-L355) in react-apollo, or [here](https://github.com/apollographql/apollo-client/blob/e6ac1840f0c57e86d5cb77c7f012526e5ec68ebb/src/react/data/QueryData.ts#L430-L454) in apollographql/apollo-client). 

I wonder: could `onError` receive `data` as a 2nd ordinal argument in the event that there is both an `error` and `data`?

## How to reproduce the issue:

I've reproduced here: https://codesandbox.io/s/inspiring-surf-pcrgj?file=/src/App.js

Here's an inline example:

```jsx
import {useQuery} from '@apollo/react-hooks`
import gql from 'graphql-tag'

const someSideEffect = (error, data) => {
if (!data) console.error(error)
 else if (data && error) {
  console.info(error)
  alert(`some of the data is shiny: ${data.isShiny}`)
 } else {
   console.log('no error 🙌')
 }
}

const QUERY = gql `
query get_stuff {
 stuff {
  isShiny
  someFinnickyFieldThatErrorsOutSometimes
 }
}
`

function Stuff() {
 const {data, loading} = useQuery(QUERY, {
  errorPolicy: 'all',
  onCompleted: (data) => someSideEffect(null, data),
  onError: (error) => someSideEffect(error),
 });

 if (loading) return 'loading...'
 if (!loading && data) return <div>{JSON.stringify(data, null, 4)}</div>;
}
```

### Possible solution 1: provide `data` as a 2nd ordinal argument to `onError`

```jsx
const {data, loading} = useQuery(QUERY, {
 errorPolicy: 'all',
 onCompleted: (data) => someSideEffect(null, data),
 // Perhaps It'd be SO nice if onError had access to data ❤, but it does not 💔
 onError: (error, data) => someSideEffect(error, data),
});
```

One might also argue additionally (or alternatively) that error in the `onCompleted` callback should be made available. Hence, in the case of errorPolicy="all" this dichotomy between "completed" and "errored" is somewhat difficult to discern. Hence:

### Possible Solution 3: Alternatively, and more generally it should just always call "onCompleted" with access to both error and data when both are defined (errorPolicy="all"). And we remove the notion of onError entirely?\*\*

In summary, there are many solutions to this perceived bug, and the API seems unclear for this use case. These three solution proposals above are ranked in order from least to most invasive.

## Version

**@apollo/react-components@3.1.5**

```bash
System:
OS: macOS Mojave 10.14.6
Binaries:
Node: 12.16.1 - ~/.volta/tools/image/node/12.16.1/6.13.4/bin/node
npm: 6.13.4 - ~/.volta/tools/image/node/12.16.1/6.13.4/bin/npm
Browsers:
Chrome: 83.0.4103.116
Edge: 83.0.478.61
Firefox: 76.0
Safari: 13.0.4
```

(FYI originally reported here, moving)
https://github.com/apollographql/react-apollo/issues/403

**Comment by llamadeus** [comment]
I'm currently having some troubles when setting `errorPolicy: 'all'` because the `onCompleted` callback is being called even if `data` is `null`. This will result in `null` being passed for the `data` argument but the function signature does not reflect this circumstance.

I'm using `"@apollo/client": "^3.2.7"`.

**Comment by jwm0** [comment]
It's not perfect but you can use `client.readQuery` inside of `onError` to read the data that was fulfilled.

**Comment by tdeedlesATX** [comment]
Thanks @jwm0 ! Your suggestion is an adequate workaround. I do not love though that this workaround takes an imperative approach to a defect in the declarative `onError`.  

But your solution is perhaps than what we currently have as workaround! Thank you ❤️

**Comment by zorji** [comment]
I am testing with `apollo/client@3.4.1`, I got a completely different result.
When `errorPolicy=all`, it does not trigger `onError` on partial failure.

I can't find the official explanation on the website, when `errorPolicy=all` and when there is a partial error, should it trigger `onError`?

I found this PR https://github.com/apollographql/apollo-client/issues/6492 and it looks like it was not expected on trigger `onError`.

**Comment by ikushlianski** [comment]
I faced the same issue with `errorPolicy: 'all'`, I only received the partial result inside onCompleted callback of useQuery, but no errors.

I created an ErrorLink like this:

```typescript
private createErrorLink = () => {
    return onError(
      ({ operation, response, forward, graphQLErrors, networkError }) => {
        return forward(operation).map((response) => {
          if (response?.data) {
            response.data.errors = graphQLErrors;
          }

          return response;
        });
      }
    );
  };
```

Now inside my onCompleted callback I get my data as well as errors. You will have to tweak Apollo types a bit, because it does not have errors field by default.

```typescript
const [getAnalyses] = useLazyQuery<
    GetAnalysesQuery & { errors: GraphQLError[] }
  >(GetAnalysesDocument, {
    variables: { environmentContexts },
    client: quicksightApolloClient,
    fetchPolicy: "no-cache",
    onCompleted: ({ getAnalyses, errors }) => {
      console.log("errors from BE", errors);
      // other logic
    },

    onError: (error) => {
      // we get here only in case of NetworkError
  });
```

**Comment by phryneas** [comment]
Hello everyone!

We just released Apollo Client 4.0, and in doing so we removed the (already deprecated for a while) `onError` and `onCompleted` callbacks from `useQuery`.

I believe in that context it's probably best to close this issue - it just doesn't apply anymore.

Have a great day!
